sgkit_vcf/vcf_reader.py

```diff
@@ -0,0 +1,286 @@
+) -> None:
+    """Convert specified regions of one or more VCF files to zarr files, then concat, rechunk, write to zarr"""
+
+    if isinstance(input, str) or isinstance(input, Path):
+        # Single input
+        inputs: Sequence[PathType] = [input]
+        assert regions is not None  # this would just be sequential case
+        input_regions: Sequence[Optional[Sequence[str]]] = [regions]  # type: ignore
+    else:
+        # Multiple inputs
+        inputs = input
+        if regions is None:
+            input_regions = [None] * len(inputs)
+        else:
+            if len(regions) == 0 or isinstance(regions[0], str):
+                raise ValueError(
+                    f"Multiple input regions must be a sequence of sequence of strings: {regions}"
+                )
+            input_regions = regions
+
+    assert len(inputs) == len(input_regions)
+
+    tempdir = Path(tempfile.mkdtemp(prefix="vcf_to_zarr_"))
+
+    datasets = []
+    parts = []
+    for i, input in enumerate(inputs):
+        filename = Path(input).name
+        input_region_list = input_regions[i] or [None]  # type: ignore
+        for r, region in enumerate(input_region_list):
+            part = tempdir / filename / f"part-{r}.zarr"
+            parts.append(part)
+            ds = dask.delayed(vcf_to_dataset)(
+                input,
+                output=part,
+                region=region,
+                chunk_length=chunk_length,
+                chunk_width=chunk_width,
+            )
+            datasets.append(ds)
+    datasets = dask.compute(*datasets)
+
+    # Ensure Dask task graph is efficient, see https://github.com/dask/dask/issues/5105
+    dask.config.set({"optimization.fuse.ave-width": 50})
+
+    # Combine the datasets into one
+    ds = xr.concat(datasets, dim="variants", data_vars="minimal")  # type: ignore[no-untyped-call, no-redef]
+    ds: xr.Dataset = ds.chunk({"variants": chunk_length, "samples": chunk_width})  # type: ignore
+
+    # Set variable length strings to fixed length ones to avoid xarray/conventions.py:188 warning
+    # (Also avoids this issue: https://github.com/pydata/xarray/issues/3476)
+    max_variant_id_length = max(ds.attrs["max_variant_id_length"] for ds in datasets)
+    max_variant_allele_length = max(
+        ds.attrs["max_variant_allele_length"] for ds in datasets
+    )
+    ds["variant_id"] = ds["variant_id"].astype(f"S{max_variant_id_length}")
+    ds["variant_allele"] = ds["variant_allele"].astype(f"S{max_variant_allele_length}")
+    del ds.attrs["max_variant_id_length"]
+    del ds.attrs["max_variant_allele_length"]
+
+    delayed = ds.to_zarr(output, mode="w", compute=False)
+    # delayed.visualize()
+    delayed.compute()
+
+    # Delete intermediate files from temporary directory
+    shutil.rmtree(tempdir)
+
+
+def vcf_to_zarr(
+    input: Union[PathType, Sequence[PathType]],
+    output: PathType,
+    *,
+    regions: Union[None, Sequence[str], Sequence[Optional[Sequence[str]]]] = None,
```

Comment:
What was the rationale for allowing multiple regions within the same file?  Is that mainly because the filter syntax only supports a single contig and there might be multiple in one file?  I was planning on mirroring this with bgen as multiple slices for a single file, but I couldn't think of a reason why it would be necessary (the extra generality doesn't hurt though).  I'd like to make sure I'm not missing an important use case for extracting non-contiguous regions of the same file.

Reply:
> Is that mainly because the filter syntax only supports a single contig and there might be multiple in one file?

Exactly that. E.g. gnomAD has multiple contigs in one large VCF file. I've seen other examples of this too.